Allow associating subscopes of any level when creating new resources

leio10 opened the pull request:
#### :tophat: What? Why?

When creating a new resource, user should be able to select any subscope of the parent participatory space, not just first level subscopes. The current plain select box should be turned into a select2 box that is able to search deeper. This apply to end users when creating proposals and to admin users when creating/editing every resource that has an scope.

#### :pushpin: Related Issues
- Fixes #1708 

#### :ghost: GIF
![](http://33.media.tumblr.com/0ddff5da60bbff3f72fb736e528d6c81/tumblr_nr2b1vOBdT1rc7zl1o4_400.gif)

deivid-rodriguez commented on decidim-core/spec/controllers/scopes_controller_spec.rb:
```diff
@@ -37,27 +38,50 @@ module Decidim
+
+      context "include root" do
+        let(:params) { { term: query, root: scopes.first, include_root: true } }
+        it { expect(results["results"].map { |r| r["text"] }).to match_array %w(Aaaa Cccc) }
```
I like to isolate this repeated logic into `RSpec` matchers. In this case, this would do:

```ruby
matcher :have_scopes do |expected|
  match do |results|
    result_texts = results["results"].map { |r| r["text"] }

    RSpec::Matchers::BuiltIn::ContainExactly.new(result_texts).matches?(expected)
  end
end
```

deivid-rodriguez replied:
You forgot to use the matcher here.

deivid-rodriguez commented:
Nice work @leio10!

While reviewing this I noticed some usability things in the scopes administration. I comment it here, if we agree we can reticket them:

* Scopes are managed as subsections of the `Settings` sections. I think they should be subsections of a new main "Scopes" section.

![mainsection](https://user-images.githubusercontent.com/2887858/29271808-a41fed3c-80fd-11e7-9d5f-84f273b69017.png)

* I took me a while to figure out where to define new subscopes for a scope (You have to click on the name of the scope, not on the `edit` icon). I would propose to remove the `show` page for scopes, and merge it with `edit`. Clicking on a scope name would take you to the scope edit page, just like it happens with participatory processes. In the new edit page you would be able to:
    * Edit the main information for the scope.
    * Click on an "Add subscope" blue button to add a subscope.
    * See the list of subscopes each with their "edit" and "remove" icons.


* In the new forms, you can see a small cross to remove the selected scope. I don't think this makes sense here, since you can only select a single scope:

![gaylord daniel and doyle](https://user-images.githubusercontent.com/2887858/29271480-ee93cd5e-80fb-11e7-8d13-cff98d2ee29a.png)

leio10 commented on decidim-core/app/controllers/decidim/scopes_controller.rb:
```diff
@@ -13,7 +13,7 @@ def search
                  current_organization.top_scopes
                end
-      root_option = if params[:include_root] && !params[:term].present?
+      root_option = if params[:include_root] && params[:include_root] != "false" && !params[:term].present?
```
Mmm, I have had the same doubt, but I think it is not the same. I want it to enter the `if` block if `params[:include_root]` has any value different from `nil` and `"false"`. The first condition excludes `nil` and `"false"` is excluded by the last one.
This case occurs when you go to https://github.com/decidim/decidim/blob/master/decidim-proposals/app/views/decidim/proposals/proposals/_filters.html.erb#L32 and change `include_root: true` with `include_root: false`. As this info goes through the querystring, `false` is received as `"false"`. Even when currently there is no such case now, it happened to me twice, so I think is better this way to avoid wasting time in the future.

Regarding https://github.com/decidim/decidim/pull/1500#issuecomment-318719458, I think I didn't understand what you were saying. The extra option that appears is a "Loading..." message, that disappears when the AJAX response is received. I didn't changed anything related, maybe there was a problem with AJAX requests on that moment and the 'extra option' didn't disappear at all.

deivid-rodriguez replied:
> Mmm, I have had the same doubt, but I think it is not the same. I want it to enter the if block if params[:include_root] has any value different from nil and "false". The first condition excludes nil and "false" is excluded by the last one.
This case occurs when you go to https://github.com/decidim/decidim/blob/master/decidim-proposals/app/views/decidim/proposals/proposals/_filters.html.erb#L32 and change include_root: true with include_root: false. As this info goes through the querystring, false is received as "false". Even when currently there is no such case now, it happened to me twice, so I think is better this way to avoid wasting time in the future.

What was I thinking? :laughing: You're absolutely right, my bad!

> Regarding #1500 (comment), I think I didn't understand what you were saying. The extra option that appears is a "Loading..." message, that disappears when the AJAX response is received. I didn't changed anything related, maybe there was a problem with AJAX requests on that moment and the 'extra option' didn't disappear at all.

No, I meant to extra search results appearing when you start filtering. Normally you expect that filtering actually reduces the number of options available. But I understand now how this makes sense since showing all subscopes of all levels is not useful (and would overflow) and having a quick selector of first level scopes without searching is nice. So forget everything I said regarding this matter, it's great as it is! :+1:

deivid-rodriguez replied:
Actually, one more thing regarding the `include_root` parameter. Since as you said, we're never passing `include_root: false`, I think it's clearer (and simpler) to explicitly check for the positive case. Would this condition be equivalent?

```diff
                else
                  current_organization.top_scopes
                end
-      root_option = if params[:include_root] && params[:include_root] != "false" && !params[:term].present?
+      root_option = if params[:include_root] == "true" && !params[:term].present?
                       if root
                         [{ id: root.id.to_s, text: root.name[I18n.locale.to_s] }]
                       else
```

leio10 commented:
Thanks for the feedback, @deivid-rodriguez!

Regarding the clearing selection cross, I can think two options and I don't have clear which is the best (and I've tested both):
- It can work without a "Global scope" option as now:
  - There is a "No value" option, the "Global scope". It appears in the prompt, so is the 'default' option.
  - When you select another option, you can only choose the "Global scope" with that awful cross.
  - The "Global scope" option is clearly differentiated from other options.
- It can work with a "Global scope" option:
  - There is no "No value" option. There wouldn't be a prompt message, since a 
  - The "Global scope" should be the selected value at the create form loading, this implies some changes.
  - The "Global scope" option is no different from other options.

This situation apply to every creation/edition form where a scope is selected, so it would be nice to find the best solution and then apply it globally.

Regarding the menu issue, I think that creating a Scopes section in the main menu gives too much importance to scopes, while I estimate that most admin users won't access to the scopes zone. Anyway, I don't like seeing the "Scope types" option below the Scopes option in the menu too. Maybe in the future, if the scopes navigation allows to browse participatory processes it would have more sense.

Regarding the other issue, we have discussed a little bit here: https://github.com/decidim/design-admin/pull/22. IMHO, merging the Edit scope page with the Browse scope page would complicate things. What do you think about adding a new icon (for example, something like http://fontawesome.io/icon/sitemap) in the scopes list that takes you to the browse page?

deivid-rodriguez commented:
> It can work with a "Global scope" option:
>   * There is no "No value" option. There wouldn't be a prompt message, since a
>   * The "Global scope" should be the selected value at the create form loading, this implies some changes.
>   * The "Global scope" option is no different from other options.

Although the first bullet point is incomplete, this seems a priori like the better option to me.

> Regarding the menu issue, I think that creating a Scopes section in the main menu gives too much importance to scopes, while I estimate that most admin users won't access to the scopes zone. Anyway, I don't like seeing the "Scope types" option below the Scopes option in the menu too. Maybe in the future, if the scopes navigation allows to browse participatory processes it would have more sense.

I can see how this makes sense, :+1:.

> Regarding the other issue, we have discussed a little bit here: decidim/design-admin#22. IMHO, merging the Edit scope page with the Browse scope page would complicate things. What do you think about adding a new icon (for example, something like http://fontawesome.io/icon/sitemap) in the scopes list that takes you to the browse page?

You mean that it would complicate things from the implementation point of view? From the user point of view, I can't think of a clear downside, since the edit page would also display all the information about the scope. The same thing happens with participatory processes, there's no show page for them. Well, there's is, but we don't use it (#1742).

leio10 commented:
> Although the first bullet point is incomplete, this seems a priori like the better option to me.

Sorry about that, I completed it later :see_no_evil:.

> You mean that it would complicate things from the implementation point of view? From the user point of view, I can't think of a clear downside, since the edit page would also display all the information about the scope. The same thing happens with participatory processes, there's no show page for them. Well, there's is, but we don't use it (#1742).

I mean that it would be complicated for the user. Merging both pages would allow to press the "Add subscope" button while having unsaved changes in the edit form. In my head, browsing scopes until reaching the desired one, and then modify it or add it a subscope seems pretty straightforward.

As I said before, an extra link to browse scope would make more explicit the browse behavior:
![imagen](https://user-images.githubusercontent.com/453545/29275423-54852dba-810b-11e7-891f-4195a704179b.png)

Anyway, both issues are kind of subjective, it would be nice to create a new ticket to reach a broadly accepted approach as you said.

deivid-rodriguez commented:
> I mean that it would be complicated for the user. Merging both pages would allow to press the "Add subscope" button while having unsaved changes in the edit form. In my head, browsing scopes until reaching the desired one, and then modify it or add it a subscope seems pretty straightforward.

You mean the user would click on "Add subcope" thinking that's the submit button for the form? The blue top action buttons in the admin never submit forms if IIRC, but I guess it could confuse first time users (specially if not worded properly).

I still think merging edit & show is better, but I agree it's something subjective. I conceptually see both changing the name and adding subscopes as "editing the scope", that's why I kept on clicking edit and couldn't find the way of adding a scope. That said, the new icon would have certainly saved me most of the confusion, so I'm fine with that option as well, :+1:. Let's wait for input from others!

deivid-rodriguez commented:
Reticketed as #1747 and #1748.

I offered an alternative in #1748 (nested forms) to avoid hitting the "Add subscope" confusion you mentioned.

leio10 commented:
@deivid-rodriguez  Thanks for all the help! :heart: :heart:
I've fixed the pending test and added the browse icon to leave this PR ready to merge. :smiley:

deivid-rodriguez commented:
@leio10 So sorry but now (after my own matcher suggestion :sweat_smile:) the specs read a bit weird, I would add this last improvement:

```diff
@@ -17,6 +17,8 @@ module Decidim
     let(:params) { { term: query } }
     let(:results) { JSON.parse(response.body) }
 
+    subject { results["results"] }
+
     before do
       @request.env["decidim.current_organization"] = organization
       sign_in user, scope: :user
@@ -25,7 +27,7 @@ module Decidim
 
     matcher :have_scopes do |expected|
       match do |results|
-        result_texts = results["results"].map { |r| r["text"] }
+        result_texts = results.map { |r| r["text"] }
 
         RSpec::Matchers::BuiltIn::ContainExactly.new(result_texts).matches?(expected)
       end
@@ -37,58 +39,58 @@ module Decidim
       end
 
       it "result has id" do
-        expect(results["results"].first).to have_key("id")
+        expect(subject.first).to have_key("id")
       end
 
       it "result has text" do
-        expect(results["results"].first).to have_key("text")
+        expect(subject.first).to have_key("text")
       end
     end
 
     context "search top scopes" do
-      it { expect(results).to have_scopes %w(Aaaa Aabb Bbbb) }
+      it { is_expected.to have_scopes %w(Aaaa Aabb Bbbb) }
     end
 
     context "find one result" do
       let(:query) { "Bb" }
-      it { expect(results).to have_scopes %w(Bbbb) }
+      it { is_expected.to have_scopes %w(Bbbb) }
     end
 
     context "find several results" do
       let(:query) { "Aa" }
-      it { expect(results).to have_scopes %w(Aaaa Aabb) }
+      it { is_expected.to have_scopes %w(Aaaa Aabb) }
     end
 
     context "find subscopes" do
       let(:query) { "Cc" }
-      it { expect(results).to have_scopes %w(Cccc) }
+      it { is_expected.to have_scopes %w(Cccc) }
     end
 
     context "don't find results" do
       let(:query) { "Dd" }
-      it { expect(results["results"]).to be_empty }
+      it { is_expected.to be_empty }
     end
 
     context "with root filter" do
       let(:params) { { term: query, root: scopes.first } }
 
       context "search top scopes" do
-        it { expect(results).to have_scopes %w(Cccc) }
+        it { is_expected.to have_scopes %w(Cccc) }
       end
 
       context "find one result" do
         let(:query) { "Cc" }
-        it { expect(results).to have_scopes %w(Cccc) }
+        it { is_expected.to have_scopes %w(Cccc) }
       end
 
       context "don't find results outside the root scope" do
         let(:query) { "Bb" }
-        it { expect(results["results"]).to be_empty }
+        it { is_expected.to be_empty }
       end
 
       context "include root" do
         let(:params) { { term: query, root: scopes.first, include_root: true } }
-        it { expect(results).to have_scopes %w(Aaaa Cccc) }
+        it { is_expected.to have_scopes %w(Aaaa Cccc) }
       end
     end
   end
```

I tried to add it myself to not bother you anymore, but I don't have permissions to push to your fork... :pray:.

leio10 commented:
@deivid-rodriguez Don't worry, I enjoy learning this :)

josepjaume commented:
This is awesome! Merging it! Thank you all for the good work! ❤️
